src/cpp/core/FileSerializer.cpp

```diff
@@ -24,13 +24,62 @@
 #include <boost/algorithm/string/trim.hpp>
 #include <boost/algorithm/string/predicate.hpp>
 #include <boost/iostreams/copy.hpp>
+#include <boost/thread.hpp>
 
 #include <shared_core/FilePath.hpp>
+#include <core/DateTime.hpp>
 #include <core/StringUtils.hpp>
 
 namespace rstudio {
 namespace core {
 
+namespace {
+
+bool isFileLockedError(const Error& error)
+{
+   // exclusive file access is only present on Windows
+#ifndef _WIN32
+   return false;
+#endif
+   return (error && error.getCode() == ERROR_SHARING_VIOLATION);
+}
+
+Error openFileForWritingWithRetry(const FilePath& filePath,
+                                  bool truncate,
+                                  int maxOpenRetrySeconds,
+                                  std::shared_ptr<std::ostream>* pOfs)
+{
+   using namespace boost::posix_time;
+
+   ptime startTime = second_clock::universal_time();
+   Error lastError;
+
+   // do not allow negative values - regular signed int was chosen here for
+   // easier integration with other parts of the codebase
+   if (maxOpenRetrySeconds < 0)
+      maxOpenRetrySeconds = 0;
+
+   while (true)
+   {
+      lastError = filePath.openForWrite(*pOfs, truncate);
+
+      // if the error is a non file lock error, then we should just return it
+      if (!isFileLockedError(lastError))
+         return lastError;
+
+      // stop retrying if we've spent more than the requested amount of time
+      if ((second_clock::universal_time() - startTime) >= seconds(maxOpenRetrySeconds))
+         break;
```

Comment:
Can we amend the primary error object here with a property that indicates that we timed out (and possibly how many tries/attempts we made)? I also think this error would be worth logging _and_ returning since a save failure can cause data loss.